File: Esame6Java/app/src/main/java/com/example/esame6java/MainActivity.java
package com.example.esame6java;

import android.app.Activity;
import android.app.AlertDialog;
import android.content.DialogInterface;
import android.os.Bundle;
import android.view.View;
import android.widget.AdapterView;
import android.widget.EditText;
import android.widget.ListView;
import android.widget.Toast;

import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends Activity {

    EditText etTitolo;
    EditText etAutore;
    ListView lvLibro;

    List<Libro> libri = new ArrayList<>();
    CustomAdapter adapter;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        etTitolo = findViewById(R.id.etTitolo);
        etAutore = findViewById(R.id.etAutore);
        lvLibro = findViewById(R.id.lvLibro);

        libri.add(new Libro("The Hobbit", "Tolkien"));
        adapter = new CustomAdapter(this, R.layout.book_element, libri);
        lvLibro.setAdapter(adapter);

        lvLibro.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                showDeleteDialog(position);
            }
        });
    }

    private void showDeleteDialog(final int position) {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setTitle("Delete Item");
        builder.setMessage("Are you sure you want to delete this item?");

        // Set up the "Yes" button
        builder.setPositiveButton("Yes", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {

                libri.remove(position);

                adapter.notifyDataSetChanged();


                Toast.makeText(MainActivity.this, "Item deleted", Toast.LENGTH_SHORT).show();
            }
        });


        builder.setNegativeButton("No", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {

                dialog.dismiss();
            }
        });


        builder.create().show();
    }

    public void inserisciLibro(View v){
        String titolo = etTitolo.getText().toString();
        String autore = etAutore.getText().toString();

        if(titolo.strip().equals("") || autore.strip().equals("")){
            Toast.makeText(getApplicationContext(), "Input incompleto", Toast.LENGTH_SHORT).show();
        } else{
            Libro l = new Libro(titolo, autore);
            libri.add(l);
            adapter.notifyDataSetChanged();
        }
    }


}

File: Esame4Kotlin/app/src/main/java/com/example/esameguessword/MainActivity.kt
package com.example.esameguessword

import android.os.Bundle
import android.view.View
import android.widget.EditText
import android.widget.TextView
import android.widget.Toast
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import android.content.DialogInterface
import kotlin.random.Random


class MainActivity : AppCompatActivity() {

    val words = arrayOf("GUESS", "RANDOM", "WORD", "NOW","RAINBOW","WONDER", "MOBILE", "UNISA", "INSTAGRAM",
    "ESAME", "ANDROID")
    lateinit var tvWord : TextView
    lateinit var etGuess : EditText
    lateinit var tvCount : TextView
    lateinit var displayWord : CharArray
    lateinit var randomWord : String
    var countTry = 0
    var isCompleted = false


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        randomWord = words.random()
        displayWord = "_".repeat(randomWord.length).toCharArray()

        tvWord = findViewById(R.id.tvWord)
        etGuess = findViewById(R.id.etGuess)
        tvCount = findViewById(R.id.tvCount)

        updateword()

    }

    fun updateword(){
        if(!isCompleted){
        tvWord.text = displayWord.joinToString(" ")
        tvCount.text = "tentativi = $countTry"
        }
        if(displayWord.joinToString("") == randomWord){
            Toast.makeText(this, "Guessed Correctly", Toast.LENGTH_SHORT).show()
            isCompleted = true
        }
    }

    fun guessLetter(v : View){

        if(etGuess.text.isNotEmpty()){
            val letGuess = (etGuess.text.toString()).uppercase()[0]
            if(letGuess in randomWord && letGuess !in displayWord){
                val indexes : List<Int> = randomWord.mapIndexedNotNull { index, c ->
                    if (c == letGuess) index else null
                }
                for(index in indexes){
                    displayWord[index] = letGuess
                }
            }else{
                countTry++
            }
            updateword()
        }
    }

    fun resetGame(v: View){
        MaterialAlertDialogBuilder(this)
            .setTitle("Reset")
            .setMessage("Are you sure you want to reset")
            .setNegativeButton("No"
            ) { dialog, whichButton ->
                Toast.makeText(this, "Not Reseted", Toast.LENGTH_SHORT).show()
            }
            .setPositiveButton("Yes"
            ) { dialog, whichButton ->
                val oldWord = randomWord
                while(oldWord == randomWord) {
                    randomWord = words.random()
                }
                isCompleted = false
                countTry = 0
                displayWord = "_".repeat(randomWord.length).toCharArray()
                updateword()
            }
            .show()
    }

    fun helpGame(v:View){
        if(isCompleted){
            Toast.makeText(this, "Word is already completed", Toast.LENGTH_SHORT).show()
        } else{
            countTry += 5
            var randomLet = Random.nextInt(randomWord.length)
            while (displayWord[randomLet] != '_'){
                randomLet = Random.nextInt(randomWord.length)
            }
            displayWord[randomLet] = randomWord[randomLet]
            updateword()
        }

    }




}